(refactor): 2053 nnnnat address book to react

nnnnat opened the pull request:
Resolves #2053   
Impact: **breaking**  
Type: **refactor**

## Issue
The AddressBook components were still using Blaze templates.

## Solution
Refactored the blaze AddressBook components to React
 * Created a new AddressBookContainer that binds data and reducers to the component.
 * Created a new AddressBook component that creates the base layout and handles data changes
 * Created a new AddressBookForm component that handles adding/editing an address.
 * Created a new AddressBookGrid component that displays the shipping/billing address.
 * Updated the Profile and Checkout blaze templates to use the new react AddressBook

## Breaking changes
The new React components should be working the same the old ones, but they are no longer blaze templates. 

All of the new components are available from the `@reactioncommerce/reaction-components` package.

## Testing
The AddressBook needs to be tested from both the Profile and Checkout views.

### Setup
**Profile**
1. Log in as an Admin or create a new account.
2. From the drop-down click the profile link.
3. Follow the testing steps below.

**Checkout**
1. As an anonymous customer, add a few items to the Cart and go to check out.
2. Enter an email address.
3. Follow the testing steps below.

### Address Book Test
**Adding a new address**
1. Enter a new shipping/billing address.
2. Click the "save and continue" button.
3. The new address should now display in the AddressBookGrid and checkout step 3 should be available.

**Adding a second address**
1. Now click the "Add Address" button at the top of the AddressBook
2. Enter the second address.
3. Both addresses should display in the AddressBookGrid.

**Changing Shipping/Billing address**
1. Click on the new address to make it the default billing address
2. The blue checkmark should show up on the new address and the shipping should recalculate.
3. Try switching the default billing/shipping address using the keyboard.

**Editing an address**
1. Click the edit button of an address in the AddressBookGrid.
2. Change the some of the fields and click "save and continue".
3. The updated address should now show in AddressBookGrid.

**Removing an address**
1. Click the delete/trash button of an address in the AddressBookGrid.
2. Address should now be removed from the grid.

## Notes
### Meteor methods
I created some helper/reducer methods for the AddressBook inside the AddressBookContainer. I'm just wrapping the Meteor methods that do the CRUD operations on the account's addresBook. I've copy most of that code out of the template helpers and I'm not confident I'm handling the callbacks correctly.

### Validation
**General Validation**
Currently, this version of the AddressBook only shows a toast alert if there is an address validation issue, I know we have a few issues around form/address validation being worked on and I wanted to get this reviewed so others can weigh in on how to best handle displaying field errors.

**Avalara Validation and the AddressReview view**
While testing with Avalara enabled I was unable to get the AddressReview template to show even if Avalara threw an error on the server, the invalid address would be added to the address grid anyway.
Could be related to issue #3518 could be user error.

I left the blaze addressBook template code in the project, for now, in case it's needed as a reference but once this PR is approved I'll remove those files.

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,131 @@
+
+    updateAddress = (address, property) => updateAddress(address, property);
+
+    removeAddress = (_id) => removeAddress(_id);
```
These three lines aren't needed. Just remove `this.` below and pass the three functions directly.

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,131 @@
+  } else {
+    Meteor.call("accounts/addressBookUpdate", address);
+  }
+}
```
You should deal with any potential errors by having `Meteor.call` callbacks like the other two functions have. But also, I recommend making all of these return promises, which reject in case of error. Then, rather than calling `Alerts.toast` directly in these action functions, you should supply an `onError` function to all of the components.

So for example removeAddress becomes:

```js
function removeAddress(_id) {
  return new Promise((resolve, reject) => {
    Meteor.call("accounts/addressBookRemove", _id, (error, result) => {
      if (error || !result) {
        reject(i18next.t("addressBookGrid.cantRemoveThisAddress", { err: error && error.message }));
      } else {
        resolve(result);
      }
    });
  });
}
```

And then in your component you do

```js
onRemove = (_id) => {
  const { onError, removeAddress } = this.props;
  removeAddress(_id).catch(onError);
}
```

passing an `onError` similar to:

```js
function onError(errorMessage) {
  Alerts.toast(errorMessage, "error");
}
```

This keeps better separation which will make testing and debugging easier.

aldeed commented on imports/plugins/core/checkout/client/templates/checkout/addressBook/addressBook.js:
```diff
@@ -0,0 +1,28 @@
+   * @return {Object} - contains the component for updating a user's address.
+   */
+  AddressBook() {
+    const { status, position } = Template.instance().data;
```
`template.data` is the nonreactive one. Use `Template.currentData()` instead to ensure it will update when `status` or `position` change after initial render.

```js
const { status, position } = Template.currentData();
```

aldeed commented:
@nnnnat Overall this looks great, and was clearly a lot of work. Nice job! I suggested some changes to try to better separate certain things, deal with edge cases, and improve prop types.

nnnnat commented:
@aldeed I updated this PR with the requested changes and it's ready for review.

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,214 @@
+    }
+    country[key] = regions;
+    regionsByCountry = { ...regionsByCountry, ...country };
+  });
```
This will create a new map every loop, which is inefficient. Since you're just building the `regionsByCountry` map right now, you don't have to worry about mutating it causing problems. So you can do:

```js
const regionsByCountry = {};
```

And then skip the `country` variable and instead directly do `regionsByCountry[key] = regions;`

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,214 @@
+
+  let regionsByCountry;
+  Object.keys(shopCountries).forEach((key) => {
+    const { states } = shopCountries[key] || undefined;
```
The end of this should be `|| {}` since you can't destructure `undefined`.

aldeed commented on imports/plugins/core/accounts/client/components/addressBook.js:
```diff
@@ -0,0 +1,397 @@
+        label: PropTypes.String,
+        value: PropTypes.String
+      }))
+    }),
```
This prop type is wrong. There isn't a key named "countryCode" but rather one arbitrary key for each country code. So since this is a map, it's OK to do `regionsByCountry: PropTypes.object` instead of shape.

aldeed commented on imports/plugins/core/accounts/client/components/addressBookForm.js:
```diff
@@ -0,0 +1,336 @@
+        label: PropTypes.String,
+        value: PropTypes.String
+      }))
+    })
```
See other comment. `regionsByCountry: PropTypes.object`

aldeed commented on imports/plugins/core/accounts/client/components/addressBookGrid.js:
```diff
@@ -0,0 +1,224 @@
+   * @return {String} - default shipping address _id or empty string.
+   */
+  get defaultShippingAddressId() {
+    const { addressBook } = this.props;
```
Since you're calling `.find` blindly on this, I recommend the same `this.addressBook` getter pattern you did in the other component.

aldeed commented on imports/plugins/core/accounts/client/components/addressBookGrid.js:
```diff
@@ -0,0 +1,224 @@
+   * @return {String} - default billing address _id or empty string.
+   */
+  get defaultBillingAddressId() {
+    const { addressBook } = this.props;
```
Since you're calling `.find` blindly on this, I recommend the same `this.addressBook` getter pattern you did in the other component.

aldeed commented on imports/plugins/core/accounts/client/components/addressBookGrid.js:
```diff
@@ -0,0 +1,224 @@
+     * selects a default shipping or billing address
+     */
+    select: PropTypes.func
+  }
```
Add default no-ops for all functions since you call them without checking whether they're set.

```js
static defaultProps = {
  edit() {},
  remove() {},
  select() {}
};
```

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,214 @@
+  return new Promise((resolve, reject) => {
+    Meteor.call("accounts/validateAddress", address, (error, result) => {
+      if (error || !result) reject(i18next.t("addressBookAdd.failedToAddAddress", { err: error.message }));
+      if (result && result.validated) {
```
This second `if` leaves the possibility of a swallowed error, when `result.validated` is falsy. I think it'd be better to remove this second `if` and change above to `if (error || !result || !result.validated)`

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,214 @@
+    if (property) {
+      Meteor.call("accounts/addressBookUpdate", address, null, property, (error, result) => {
+        if (error || !result) reject(i18next.t("addressBookGrid.somethingWentWrong", { err: error.message }));
+        if (result) resolve(result);
```
If for some reason `error` and `result` are ever both truthy, this would end up calling `reject` and `resolve`, which results in an error. It's best to use `else` for all of these situations to ensure it will never call both.

```js
if (error || !result) {
  reject(i18next.t("addressBookGrid.somethingWentWrong", { err: error.message }));
} else {
  resolve(result);
}
```

With promises we always want to make sure exactly one of resolve or reject will be called eventually, down every possible path.

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,214 @@
+        if (error || !result) reject(i18next.t("addressBookGrid.somethingWentWrong", { err: error.message }));
+        if (result) resolve(result);
+      });
+    } else {
```
I'm not entirely sure that the `if (property)` check and this `else` are necessary. If `property` is falsy, then it will just send method arguments `address, null, null`, which will be treated the same way, looking at the server method code.

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,212 @@
+          value: PropTypes.String
+        }))
+      })
+    }
```
This `regionsByCountry` prop type should be fixed, too. However, because this is a very lightweight wrapper, I'd suggest not setting the explicit propTypes here but rather grabbing them off Comp.

```js
static propTypes = Comp.propTypes
```

aldeed commented on imports/plugins/core/accounts/client/containers/addressBookContainer.js:
```diff
@@ -0,0 +1,212 @@
+        } else {
+          resolve(res);
+        }
+      });
```
`reject` does not return out of the function, so for errors in the outer call it will still call `accounts/addressBookAdd` with this logic. You can just add an explicit return:

```js
if (error || !result || !result.validated) {
  reject(i18next.t("addressBookAdd.failedToAddAddress", { err: error.message }));
  return;
}
```

nnnnat commented:
@aldeed I've made the requested changes and fixed a few issues I caused when making changes earlier today. Everything should be working again and I believe I've got all the requested changes in. I was having one issue with the updateAddress reducer, I refactored it like you suggested removing the `if (property)` check to something like this

```js
function updateAddress(address, property) {
  return new Promise((resolve, reject) => {
      Meteor.call("accounts/addressBookUpdate", address, null, property, (error, result) => {
        if (error || !result) {
          reject(i18next.t("addressBookGrid.somethingWentWrong", { err: error.message }));
        } else {
          resolve(result);
        }
      });
  });
}
```
updating the default shipping/billing worked but updating the address fields would give me this Match error from the server
```sh
Exception while invoking method 'accounts/addressBookUpdate' Error: Match error: Failed Match.OneOf, Match.Maybe or Match.Optional validation
    at check (packages/check/match.js:35:15)
    at MethodInvocation.addressBookUpdate (server/methods/accounts/accounts.js:419:3)
    at packages/check/match.js:108:16
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at Object._failIfArgumentsAreNotAllChecked (packages/check/match.js:107:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:719:19)
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at DDPServer._CurrentWriteFence.withValue (packages/ddp-server/livedata_server.js:717:46)
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at Promise (packages/ddp-server/livedata_server.js:715:46)
    at new Promise (<anonymous>)
    at Session.method (packages/ddp-server/livedata_server.js:689:23)
    at packages/ddp-server/livedata_server.js:559:43
Sanitized and reported to the client as: Match failed [400]
```
I tried updating the params to be `function updateAddress(address, property = null)` and still got the match error. I reverted it to use the `if (propery)` and call two different update meteor methods since that seems to be working.

nnnnat commented:
@aldeed  I've updated the `accounts/addressBookUpdate` Meteor method to do `check(type, Match.Maybe(String)` and I've refactored the `updateAddress` reducer to have only one Meteor method call. The AddressBook seems to be working in both views and updating the address from both the grid and form is working.

spencern commented:
Closing in favor of #4054 which will include the work done here as well as the address validation work.
